javaslang/src/main/java/javaslang/collection/RedBlackTree.java

```diff
@@ -412,276 +466,538 @@ public boolean contains(T value) {
-                                        empty);
-                                final Node<T> newRight = new Node<>(BLACK, blackHeight, ln.right, value, right, empty);
-                                return new Node<>(RED, blackHeight + 1, newLeft, ln.value, newRight, empty);
-                            }
-                        }
-                        if (!ln.right.isEmpty()) {
-                            final Node<T> lrn = (Node<T>) ln.right;
-                            if (lrn.color == RED) {
-                                final Node<T> newLeft = new Node<>(BLACK, blackHeight, ln.left, ln.value, lrn.left,
-                                        empty);
-                                final Node<T> newRight = new Node<>(BLACK, blackHeight, lrn.right, value, right, empty);
-                                return new Node<>(RED, blackHeight + 1, newLeft, lrn.value, newRight, empty);
-                            }
-                        }
-                    }
-                }
+        public Option<Node<T>> maxNode() {
+            Node<T> curr = this;
+            while (!curr.right().isEmpty()) {
+                curr = curr.right();
             }
-            return new Node<>(color, blackHeight, left, value, right, empty);
+            return Option.some(curr);
         }
 
-        private static <T> Node<T> balanceRight(Color color, int blackHeight, RedBlackTree<T> left, T value,
-                                                RedBlackTree<T> right, Empty<T> empty) {
-            if (color == BLACK) {
-                if (!right.isEmpty()) {
-                    final Node<T> rn = (Node<T>) right;
-                    if (rn.color == RED) {
-                        if (!rn.right.isEmpty()) {
-                            final Node<T> rrn = (Node<T>) rn.right;
-                            if (rrn.color == RED) {
-                                final Node<T> newLeft = new Node<>(BLACK, blackHeight, left, value, rn.left, empty);
-                                final Node<T> newRight = new Node<>(BLACK, blackHeight, rrn.left, rrn.value, rrn.right,
-                                        empty);
-                                return new Node<>(RED, blackHeight + 1, newLeft, rn.value, newRight, empty);
-                            }
-                        }
-                        if (!rn.left.isEmpty()) {
-                            final Node<T> rln = (Node<T>) rn.left;
-                            if (rln.color == RED) {
-                                final Node<T> newLeft = new Node<>(BLACK, blackHeight, left, value, rln.left, empty);
-                                final Node<T> newRight = new Node<>(BLACK, blackHeight, rln.right, rn.value, rn.right,
-                                        empty);
-                                return new Node<>(RED, blackHeight + 1, newLeft, rln.value, newRight, empty);
-                            }
-                        }
-                    }
-                }
```

Comment:
Understood.

The idea was to make them polymorphic and avoid casting from `RedBlackTree` to `ValueNode` all the time (performance enhancement + cleaner code). Also, after creating the `AbstractSubTreeView` for the sub views, I had to move the node specific methods from `RedBlackTree` into the `Node` interface, so it made little sense to keep those static operations in `RedBlackTree` anyway.

I didn't change the logic in those static methods. My original idea was to turn them into instance methods. But the logic inside those methods is too complicated so I decided to leave it for another PR. The only changes I made was to use the more specific `ValueNode` interface instead of `RedBlackTree` to avoid casting, and let IntelliJ automatic refactoring tidy up here and there.